feat(blog): Infrastructure Alerts with PagerDuty (#268)
someengineering/some.engineering: blog/2023/02-28-pagerduty-infrastructure-alerts/index.md

This post is not intended to show how a security baseline is established - that's a much bigger topic we can not tackle in one post. For the sake of an example, we pick one common rule in a security baseline to show how we can check our infrastructure and how we can react if this baseline is violated.

## Example: Restrict Network Ingress Traffic

> A network Access Control List (ACL) rules determine what traffic is allowed or denied in a network. It is typically implemented at the router or firewall level to filter traffic based on criteria such as source and destination IP addresses, ports, and protocols. Our example baseline defines that ingress traffic should be restricted to a range of allowed IPv4 or IPv6 addresses to reduce the possible attack vector.

Reviewer: ```suggestion
A network Access Control List (ACL) rule determines what traffic is allowed or denied in a network. It is typically implemented at the router or firewall level to filter traffic based on criteria such as source and destination IP addresses, ports, and protocols. Our example baseline defines that ingress traffic should be restricted to a range of allowed IPv4 or IPv6 addresses to reduce the possible attack vector.
```

To make this example specific, we look at AWS EC2 security groups that control [access to a VPC](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-network-acls.html). The data that is captured in Resoto can be looked up in our [Reference Documentation for `aws_ec2_security_group`](https://resoto.com/docs/reference/data-models/aws#aws_ec2_security_group). We can see that the security group can define multiple IP permissions for ingress and egress, which itself is defined with multiple IP ranges. The following Resoto search reveals all security groups that violate the rule by defining at least one permission set for ingress with:

- defining no network protocol
- allowing access from any IPv4 address (equals to IPv4 CIDR `0.0.0.0/0`) or any IPv6 address (equals to IPv6 CIDR `::/0`) in any of the defined ip ranges

```bash
> search is(aws_ec2_security_group) and group_ip_permissions[*].{ip_protocol="-1" and (ip_ranges[*].cidr_ip="0.0.0.0/0" or ipv6_ranges[*].cidr_ipv6="::/0")}
```

> Note: Usually, you define your security baseline for production accounts, while we have more relaxed requirements for Dev. It is possible to restrict our search filter to only filter the relevant accounts and regions.

Reviewer: ```suggestion
Note: Usually, you define your security baseline for production accounts, while having more relaxed requirements for development accounts. It is possible to restrict our search filter to only filter the relevant accounts and regions.
```

## Create Alerts with PagerDuty

> Once resources have been detected violating our security baseline, we want to know as soon as possible and fix them. Resoto allows for a lot of possible actions. In this setup, we want to trigger an alert and make people aware of the issue. This is where PagerDuty comes in. PagerDuty is an incident management platform that helps teams manage, respond to, and resolve critical issues quickly and efficiently. You should use PagerDuty to ensure that your team is able to respond to incidents in a timely and coordinated manner, minimizing downtime and customer impact.

Reviewer: ```suggestion
Once resources violating our security baseline have been detected, we want to know as soon as possible and fix them. Resoto allows for a lot of possible actions. In this setup, we want to trigger an alert and make people aware of the issue. This is where PagerDuty comes in. PagerDuty is an incident management platform that helps teams manage, respond to, and resolve critical issues quickly and efficiently. You should use PagerDuty to ensure that your team is able to respond to incidents in a timely and coordinated manner, minimizing downtime and customer impact.
```

> Resoto comes with a `pagerduty` command that takes the result of a search filter and creates an alert in PagerDuty. Setting up this integration is straightforward and explained in this [How-To](https://resoto.com/docs/how-to-guides/alerting/create-pagerduty-alerts) guide. The `pagerduty` command has a couple of options to define how the alert should look like. Once we followed the How-To guide, only two parameters needed to be provided, the rest is optional:

Reviewer: ```suggestion
Resoto comes with a `pagerduty` command that takes the result of a search filter and creates an alert in PagerDuty. Setting up this integration is straightforward and explained in this [How-To](https://resoto.com/docs/how-to-guides/alerting/create-pagerduty-alerts) guide. The `pagerduty` command has a couple of options to define how the alert should look like. Once we followed the How-To guide, only two parameters need to be provided, the rest is optional:
```
